### PR TITLE
Add proper scalar checks to functions bound by nn.yaml.

### PR DESCRIPTION
**Pull request by gchanan**
By default, the forward functions use the default ATen scalar checks and the backward functions
use x_->isScalar() for grad_x (with grad_input mapping to self).

These can also be overridden by specifying a dict of arg_name -> scalar_check.

If the argument is not overridden and the default mapping cannot work (because x for grad_x is not
passed to the backward), an error is raised and the scalar_check must be explicitly specified.

### CODE REVIEW

File: aten/src/ATen/nn.yaml [hunk 1]
```diff
@@ -2,91 +2,142 @@
...
   buffers: [total_weight]
+  scalar_check:
+    output: 'true'
+    total_weight: 'true'
 
 - name: nll_loss2d(Tensor self, LongTensor target, Tensor weight={}, bool size_average=true, int64_t ignore_index=-100, bool reduce=True)
   cname: SpatialClassNLLCriterion
   buffers: [total_weight]
+  scalar_check:
+    output: 'true'
```
Comment (apaszke) [inline aten/src/ATen/nn.yaml]: If this says output is always a scalar then it's wrong, because it's controlled by the `reduce` argument

Comment (gchanan) [reply aten/src/ATen/nn.yaml]: Ah ya, forgot we added that, let me fix.

Comment (gchanan) [reply aten/src/ATen/nn.yaml]: should be fixed now.

**Review by colesbury — APPROVED** [review]

**Pull request merged**